homeassistant/components/saj/sensor.py

```diff
@@ -0,0 +1,198 @@
+"""SAJ solar inverter interface."""
+import asyncio
+from datetime import date, timedelta
+import logging
+
+import pysaj
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import (
+    CONF_HOST,
+    DEVICE_CLASS_POWER,
+    DEVICE_CLASS_TEMPERATURE,
+    ENERGY_KILO_WATT_HOUR,
+    POWER_WATT,
+    MASS_KILOGRAMS,
+    TEMP_CELSIUS,
+    TEMP_FAHRENHEIT,
+)
+from homeassistant.core import callback, CALLBACK_TYPE
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import async_track_point_in_utc_time
+from homeassistant.util import dt as dt_util
+
+_LOGGER = logging.getLogger(__name__)
+
+MIN_INTERVAL = timedelta(seconds=5)
+MAX_INTERVAL = timedelta(minutes=5)
+
+UNIT_OF_MEASUREMENT_HOURS = "h"
+
+SAJ_UNIT_MAPPINGS = {
+    "W": POWER_WATT,
+    "kWh": ENERGY_KILO_WATT_HOUR,
+    "h": UNIT_OF_MEASUREMENT_HOURS,
+    "kg": MASS_KILOGRAMS,
+    "°C": TEMP_CELSIUS,
+    "": None,
+}
+
+PLATFORM_SCHEMA = vol.All(
+    PLATFORM_SCHEMA.extend(
+        {vol.Required(CONF_HOST): cv.string}, extra=vol.PREVENT_EXTRA
+    )
+)
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up SAJ sensors."""
+
+    # Init all sensors
+    sensor_def = pysaj.Sensors()
+
+    # Use all sensors by default
+    hass_sensors = []
+
+    for sensor in sensor_def:
+        hass_sensors.append(SAJsensor(sensor))
+
+    saj = pysaj.SAJ(config[CONF_HOST])
+
+    async_add_entities(hass_sensors)
+
+    async def async_saj(event):
+        """Update all the SAJ sensors."""
+        tasks = []
+
+        values = await saj.read(sensor_def)
+
+        for sensor in hass_sensors:
+            state_unknown = False
+            if not values:
+                # SAJ inverters are powered by DC via solar panels and thus are
+                # offline after the sun has set. If a sensor resets on a daily
+                # basis like "today_yield", this reset won't happen automatically.
+                # Code below checks if today > day when sensor was last updated
+                # and if so: set state to None.
+                # Sensors with live values like "temperature" or "current_power"
+                # will also be reset to None.
+                if (sensor.per_day_basis and date.today() > sensor.date_updated) or (
+                    not sensor.per_day_basis and not sensor.per_total_basis
+                ):
+                    state_unknown = True
+            task = sensor.async_update_values(unknown_state=state_unknown)
+            if task:
+                tasks.append(task)
+        if tasks:
+            await asyncio.wait(tasks)
+        return values
+
+    async_track_time_interval_backoff(hass, async_saj)
```

Comment:
We should save the return value of `async_track_time_interval_backoff` and schedule a listener for home assistant stop event that cancels the update.

Reply:
Implemented a listener bind to EVENT_HOMEASSISTANT_STOP